query-builder/src/main/java/com/datastax/oss/driver/internal/querybuilder/schema/DefaultCreateKeyspace.java

```diff
@@ -0,0 +1,106 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.querybuilder.schema;
+
+import com.datastax.oss.driver.api.core.CqlIdentifier;
+import com.datastax.oss.driver.api.querybuilder.schema.CreateKeyspace;
+import com.datastax.oss.driver.internal.querybuilder.ImmutableCollections;
+import com.google.common.collect.ImmutableMap;
+import java.util.Map;
+
+public class DefaultCreateKeyspace implements CreateKeyspace {
+
+  private final CqlIdentifier keyspaceName;
+  private final boolean ifNotExists;
+  private final ImmutableMap<String, Object> properties;
+
+  public DefaultCreateKeyspace(CqlIdentifier keyspaceName) {
+    this(keyspaceName, false, ImmutableMap.of());
+  }
+
+  public DefaultCreateKeyspace(
+      CqlIdentifier keyspaceName, boolean ifNotExists, ImmutableMap<String, Object> properties) {
+    this.keyspaceName = keyspaceName;
+    this.ifNotExists = ifNotExists;
+    this.properties = properties;
+  }
+
+  @Override
+  public CreateKeyspace withProperty(String name, Object value) {
+    return new DefaultCreateKeyspace(
+        keyspaceName, ifNotExists, ImmutableCollections.append(properties, name, value));
+  }
+
+  @Override
+  public CreateKeyspace ifNotExists() {
+    return new DefaultCreateKeyspace(keyspaceName, true, properties);
+  }
+
+  @SuppressWarnings("unchecked")
+  private String extractPropertyValue(Object property) {
+    StringBuilder propertyValue = new StringBuilder();
+    if (property instanceof String) {
+      propertyValue.append("'").append((String) property).append("'");
+    } else if (property instanceof Map) {
+      Map<String, Object> propertyMap = (Map<String, Object>) property;
+      boolean first = true;
+      propertyValue.append("{ ");
+      for (Map.Entry<String, Object> subProperty : propertyMap.entrySet()) {
+        if (first) {
+          first = false;
+        } else {
+          propertyValue.append(", ");
+        }
+        propertyValue
+            .append("'")
+            .append(subProperty.getKey())
+            .append("' : ")
+            .append(extractPropertyValue(subProperty.getValue()));
+      }
+      propertyValue.append(" }");
+      // parse
+    } else {
+      propertyValue.append(property);
+    }
+    return propertyValue.toString();
+  }
+
+  @Override
+  public String asCql() {
+    StringBuilder builder = new StringBuilder();
+
+    builder.append("CREATE KEYSPACE ");
+    if (ifNotExists) {
+      builder.append("IF NOT EXISTS ");
+    }
+
+    builder.append(keyspaceName.asCql(true));
+
+    boolean first = true;
+    for (Map.Entry<String, Object> property : properties.entrySet()) {
+      if (first) {
+        builder.append(" WITH ");
+        first = false;
+      } else {
+        builder.append(" AND ");
+      }
+      String value = extractPropertyValue(property.getValue());
+      builder.append(property.getKey()).append(" = ").append(value);
```

Comment:
So far I've been very terse with spaces, I never generate any unless they're strictly necessary. I don't have strong feelings but we should be consistent.

Reply:
sounds good, i'll follow that.